### PR TITLE
Add backupDb() and restoreDb() for easy rollback.

### PR DESCRIPTION
**Pull request by platfowner**
Summary:
- Add backupDb() and restoreDb() to DB() for easy rollback
- Do some code refactoring 

#289

### CODE REVIEW

File: db/index.js [hunk 1]
```diff
@@ -65,6 +69,108 @@ class DB {
...
+  }
+
+  /**
+   * Backs up database.
+   */
+  backupDb() {
+    const LOG_HEADER = 'backupDb';
+    if (!this.stateManager) {
+      logger.error(`[${LOG_HEADER}] No state manager: ${this.stateManager}`);
+      return;
```
Comment (liayoo) [inline db/index.js:137]: return false?

File: db/index.js [hunk 2]
```diff
@@ -65,6 +69,108 @@ class DB {
...
+  }
+
+  /**
+   * Restores backup database.
+   */
+  restoreDb() {
+    const LOG_HEADER = 'restoreDb';
+    if (!this.stateManager) {
+      logger.error(`[${LOG_HEADER}] No state manager: ${this.stateManager}`);
+      return;
```
Comment (liayoo) [inline db/index.js:157]: return false?

File: node/index.js [hunk 3]
```diff
@@ -309,35 +305,15 @@ class BlockchainNode {
...
-    } else {
-      if (!this.stateManager.deleteVersion(backupVersion)) {
-        logger.error(`[${LOG_HEADER}] Failed to delete version: ${backupVersion}`);
-      }
+      this.db.restoreDb();
     }
     return result;
   }
 
+
```
Comment (liayoo) [inline node/index.js:316]: extra line?

File: node/index.js [hunk 4]
```diff
@@ -309,35 +305,15 @@ class BlockchainNode {
...
 
-    const backupVersion = this.stateManager.createUniqueVersionName(
-        `${StateVersions.BACKUP}:${this.bc.lastBlockNumber()}`);
-    const backupRoot = this.stateManager.cloneVersion(this.db.stateVersion, backupVersion);
-    if (!backupRoot) {
-      return ChainUtil.logAndReturnError(
-          logger, 11, `[${LOG_HEADER}] Failed to clone state version: ${this.db.stateVersion}`,
-          0);
-    }
+    this.db.backupDb();
```
Comment (liayoo) [inline node/index.js:308]: Should we check the result of this function call and logAndReturnError() if it's false, like we did before?

File: db/index.js [hunk 5]
```diff
@@ -65,6 +69,108 @@ class DB {
...
+    if (!restoreRoot) {
+      logger.error(`[${LOG_HEADER}] Failed to clone state version: ${this.backupStateVersion}`);
+      return false;
+    }
+    if (this.stateManager.isFinalVersion(this.stateVersion)) {
+      if (!this.stateManager.finalizeVersion(restoreVersion)) {
+        logger.error(`[${LOG_HEADER}] Failed to finalize version: ${restoreVersion}`);
+      }
+    }
+    this.setStateVersion(restoreVersion, restoreRoot);
```
Comment (liayoo) [inline db/index.js:171]: Should we clean up the backup here? (deleteBackupStateVersion?)

File: node/index.js (hunk 4 above)
Comment (platfowner) [reply node/index.js]: Done, thanks!

File: db/index.js (hunk 5 above)
Comment (platfowner) [reply db/index.js:202]: Right, it's no longer useful. Done, thanks!

File: db/index.js (hunk 1 above)
Comment (platfowner) [reply db/index.js]: Done, thanks!

File: db/index.js (hunk 2 above)
Comment (platfowner) [reply db/index.js]: Done, thanks!

File: node/index.js (hunk 3 above)
Comment (platfowner) [reply node/index.js]: Removed.

**Review by liayoo — APPROVED** [review]
LGTM, thanks!

**Pull request merged**